### PR TITLE
MPICH derivatives easyblocks

### PR DESCRIPTION
**Pull request by damianam**
See https://github.com/hpcugent/easybuild-easyconfigs/issues/2334

### CODE REVIEW

**Comment by damianam** [comment]
ping

**Comment by boegel** [comment]
@damianam: thanks for the ping, step 0 is to fix the conflict with current `develop`: https://github.com/damianam/easybuild-easyblocks/pull/1

**Comment by damianam** [comment]
@boegel done

**Comment by boegel** [comment]
Jenkins: ok to test

File: easybuild/easyblocks/m/mpich.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                else:
+                    raise EasyBuildError("Both $%s and $%s set, can I overwrite $%s with $%s (%s) ?",
+                                         envvar, new_envvar, new_envvar, envvar, envvar_val)
+        env.unset_env_vars(vars_to_unset)
+
+        super(EB_MPICH, self).configure_step()
+
+    # make and make install are default
+
+    def sanity_check_step(self, additional_paths=None):
```
Comment (boegel) [inline easybuild/easyblocks/m/mpich.py]: don't use `additional_paths`, keep `custom_paths`

Comment (boegel) [inline easybuild/easyblocks/m/mpich.py]: well, remove it to `additional_dirs`?

Comment (boegel) [inline easybuild/easyblocks/m/mpich.py]: urgh, I meant *rename*

File: easybuild/easyblocks/m/mpich.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        # they should be replaced with FC/FCFLAGS instead.
+        # Additionally, there are a set of variables (FCFLAGS among them) that should not be set at configure time,
+        # or they will leak in the mpix wrappers.
+        # Specific variables to be included in the wrapper exists, but they changed between MPICH 3.1.4 and MPICH 3.2
+        # and in a typical scenario we probably don't want them.
+        env_var_tuples = [("F90", "FC"), ("F90FLAGS", "FCFLAGS")]
+        env_var_tuples += [("CFLAGS", "MPICHLIB_CFLAGS"), ("CPPFLAGS", "MPICHLIB_CPPFLAGS")]
+        env_var_tuples += [("CXXFLAGS", "MPICHLIB_CXXFLAGS"), ("FFLAGS", "MPICHLIB_FFLAGS")]
+        env_var_tuples += [("FCFLAGS", "MPICHLIB_FCFLAGS")]
+        env_var_tuples += [("LDFLAGS", "MPICHLIB_LDFLAGS"), ("LIBS", "MPICHLIB_LIBS")]
```
Comment (boegel) [inline easybuild/easyblocks/m/mpich.py]: reformat to this (one entry per line), keep them alphabetically sorted

```python
env_vars = {
    'F90': 'FC',
    'F90FLAGS': 'FCFLAGS',
    ...
    'LIBS', 'MPICHLIB_LIBS',
}

File: easybuild/easyblocks/m/mpich.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        vars_to_unset = []
+        for (envvar, new_envvar) in env_var_tuples:
```
Comment (boegel) [inline easybuild/easyblocks/m/mpich.py]: change to:

```python
for (envvar, new_envvar) in env_vars.items():
```

File: easybuild/easyblocks/m/mpich.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            self.log.info("Making sure any old installation is removed before we start the build...")
+            super(EB_MPICH, self).make_dir(self.installdir, True, dontcreateinstalldir=True)
+
+        # MPICH configure script complains when F90 or F90FLAGS are set,
+        # they should be replaced with FC/FCFLAGS instead.
+        # Additionally, there are a set of variables (FCFLAGS among them) that should not be set at configure time,
+        # or they will leak in the mpix wrappers.
+        # Specific variables to be included in the wrapper exists, but they changed between MPICH 3.1.4 and MPICH 3.2
+        # and in a typical scenario we probably don't want them.
+        env_var_tuples = [("F90", "FC"), ("F90FLAGS", "FCFLAGS")]
```
Comment (boegel) [inline easybuild/easyblocks/m/mpich.py]: `$FC` and `$FCFLAGS` are already defined by framework now, so just add this to `vars_to_unset`?

Comment (boegel) [inline easybuild/easyblocks/m/mpich.py]: well, add `F90` and `F90FLAGS` to `vars_to_unset`, of course:

```python
vars_to_unset = ['F90', 'F90FLAGS']
```

File: easybuild/easyblocks/m/mpich.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        Custom sanity check for MPICH
+        """
+        shlib_ext = get_shared_lib_ext()
+        custom_paths = {
+            'files': ['bin/%s' % x for x in ['mpicc', 'mpicxx', 'mpif77', 'mpif90' ]] +
+                     ['lib/lib%s' % y for x in ['fmpich', 'mpichcxx', 'mpichf90', 'mpich', 'mpl', 'opa']
+                                      for y in ['%s.%s' % (x, shlib_ext)]] +
+                     ((additional_paths or {'files': []}).get('files') or []),
+            'dirs': ['include'] + ((additional_paths or {'dirs': []}).get('dirs') or []),
+        }
```
Comment (boegel) [inline easybuild/easyblocks/m/mpich.py]: ok, I would refactor this to:

```python
if custom_paths is None:
    custom_paths = {}

custom_paths.setdefault('files', []).extend([...])
custom_paths.setdefault('dirs', []).extend([...])
```

File: easybuild/easyblocks/m/mpich.py (hunk 1 above)
Comment (damianam) [inline easybuild/easyblocks/m/mpich.py]: Mmmh, there are not just dirs, there can be also files. I chose `additional_paths` to make it the same as `custom_paths`.

Comment (boegel) [inline easybuild/easyblocks/m/mpich.py]: See other remark down below. :)

**Comment by boegel** [comment]
merged via #844, thanks @damianam!

**Pull request merged**